core/src/main/scala/org/apache/spark/scheduler/TaskInfo.scala

```diff
@@ -66,6 +66,13 @@ class TaskInfo(
    */
   var finishTime: Long = 0
 
+  /**
+   * Set this var when the current task killed by other attempt tasks, this happened while we
+   * set the `spark.speculation` to true. The task killed by others should not resubmit
+   * while executor lost.
+   */
+  var killedByOtherAttempt = false
```

Comment:
`TaskInfo` is a little funny since its DeveloperApi.  Its exposed to the user via SparkListener events.  If want to leave it publicly visible, then I think there should only be a getter, not a public `var`, and also we'd need to update the way its converted to / from json in `o.a.s.util.JsonProtocol.taskInfoToJson` etc.  But I think its perhaps better to just mark this as `private[scheduler]`, as I dont' think its actually important for the end user.

Reply:
actually, I had another thought -- rather than putting this info in `TaskInfo`, I think its more natural to add something totally private to `TaskSetManager`:

`private val killedByOtherAttempt: Array[Boolean]`

Reply:
Thanks to point out the `TaskInfo` is DeveloperApi, I address this as your second comment, add this totally private to TaskSetManager